sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2364,11 +2364,20 @@ class Analyzer(override val catalogManager: CatalogManager)
      * Resolve and rewrite all subqueries in an operator tree..
      */
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUpWithPruning(
-      _.containsAnyPattern(SCALAR_SUBQUERY, EXISTS_SUBQUERY, IN_SUBQUERY), ruleId) {
+      _.containsPattern(PLAN_EXPRESSION), ruleId) {
       // In case of HAVING (a filter after an aggregate) we use both the aggregate and
       // its child for resolution.
       case f @ Filter(_, a: Aggregate) if f.childrenResolved =>
         resolveSubQueries(f, Seq(a, a.child))
+      case j @ LateralJoin(left, right, _, condition) if left.resolved && !right.resolved =>
+        val newRight = resolveSubQuery(right, j.children)(LateralSubquery(_, _, right.exprId))
+          .asInstanceOf[LateralSubquery]
+        val newCond = if (newRight.resolved && condition.exists(!_.resolved)) {
+          condition.map(resolveExpressionByPlanOutput(_, newRight.plan))
```

Comment:
How about we overwrite `resolveChildren` in `LateralJoin`, so that we consider input columns from both left and right tables? Then we don't need the special logic here, and we also follow normal join behavior and fail analysis if the name is ambiguous (`t1 JOIN LATERAL t2` ON a = 1 while both t1 and t2 have column a).